ethereum/event/BREvent.c

```diff
@@ -268,11 +302,19 @@ eventHandlerStop (BREventHandler handler) {
         // Quit the thread by aborting the queue wait.
         eventQueueDequeueWaitAbort (handler->queue);
 
-        // Wait for the thread.
-        pthread_join (handler->thread, NULL);
-        // A mini-race here?
-        handler->thread = PTHREAD_NULL;
-
+        // Wait for the thread.  We must release the lock here because the `eventHandlerThread`
+        // could itself be waiting on the lock (only as part of a running, dispatched function).
+        // Once we give the lock, the dispatched function will complete, the `eventHandlerThread`
+        // will return to the EventQueue and see the above 'Wait Abort' cond-var signaled.  At
+        // that point, `eventHandlerThread` will finish and signal 'threadExit`.  The following
+        // cond_wait will complete and reaquire the lock.
+        //
+        // Note, while we wait we will not hold the lock but handle->thread will be non-NULL; so,
+        // we don't have a start/stop race.
+        //
+        pthread_cond_wait (&handler->threadExit, &handler->lock);
```

Comment:
Really down the rabbit hole now (threading sucks).

Since `pthread_cond_wait` drops the lock, could we have multiple callers of eventHandlerStop waiting in the call? Bunch of issues could be unearthed by that (only one is signalled so other block indefinitely, even if all signalled multiple calls to detach, etc.).

Since the `eventHandlerThread` routine clears the `handler->thread` to `PTHREAD_NULL`, I think an `eventHandlerStart` call could execute before the `pthread_cond_wait` wakes up. That might lead to logic issues with dropped events/abort resets?

Lastly, probably unrealistic but if `eventHandlerThread` ever finished “early” (i.e. while `eventHandlerStop` wasn’t waiting, like if `eventQueueDequeueWait` returning an error condition), `eventHandlerStop` would block indefinitely, I think?